### PR TITLE
UCP: add Validate requests.storage field of PD/TiKV

### PR DESCRIPTION
**Pull request by shonge**
### What problem does this PR solve? 
UCP: #1460 
Close: https://github.com/pingcap/tidb-operator/issues/1460
### What is changed and how does it work?
add validate code in pkg/apis/pingcap/v1alpha1/validation/validation.go
### Check List 

Tests 

 - Unit test
 - E2E test
 - Stability test

Code changes

 - Has Go code change

Side effects

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

**Comment by Yisaer** [comment]
I think adding the e2e cases to verify the validation and defaulting is necessary.

**Comment by cofyc** [comment]
> I think adding the e2e cases to verify the validation and defaulting is necessary.

actually not necessary, because validation functions are simple functions that take into plain objects and returns `field.ErrorList`. unit tests are enough.

**Comment by cofyc** [comment]
hi, @shonge 
you can add some unit tests in [`validation_test.go`](https://github.com/pingcap/tidb-operator/blob/master/pkg/apis/pingcap/v1alpha1/validation/validation_test.go), it already has some examples.

you can also learn how to write validation tests from [Kubernetes code](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation_test.go) which contains more examples.

**Comment by Yisaer** [comment]
The rest LGTM. Adding some unit tests would be better.

**Comment by shonge** [comment]
@aylei PTAL

File: pkg/apis/pingcap/v1alpha1/validation/validation_test.go [hunk 1]
```diff
@@ -179,3 +181,109 @@ func TestValidateAnnotations(t *testing.T) {
...
+						},
+					},
+					TiDB: v1alpha1.TiDBSpec{
+						BaseImage: "pingcap/tidb",
+						Config:    &v1alpha1.TiDBConfig{},
+					},
+				},
+			},
+		},
+	}
```
Comment (Yisaer) [inline pkg/apis/pingcap/v1alpha1/validation/validation_test.go:223]: Almost LGTM, but the unit test seems a bit of weird. I recomended to write unit test in table-driven style. FYI https://github.com/golang/go/wiki/TableDrivenTests#parallel-testing

For example, we could define testcase to determine whether the tidbcluster is valid or not, if not, which compont is going to be valid.

Here is one example. https://github.com/pingcap/tidb-operator/blob/723f7c3adda243246faac3790f917bd93b25e00e/pkg/autoscaler/autoscaler/util_test.go#L29-L94

Comment (shonge) [reply pkg/apis/pingcap/v1alpha1/validation/validation_test.go:223]: haha, I will fix it

Comment (shonge) [reply pkg/apis/pingcap/v1alpha1/validation/validation_test.go]: Resolved

**Comment by shonge** [comment]
> Almost LGTM

PTAL

**Comment by shonge** [comment]
@Yisaer PTAL

**Comment by shonge** [comment]
@cofyc PTAL

File: pkg/apis/pingcap/v1alpha1/validation/validation.go [hunk 2]
```diff
@@ -168,6 +171,15 @@ func validateComponentSpec(spec *v1alpha1.ComponentSpec, fldPath *field.Path) fi
 	return allErrs
 }
 
+//validateRequestsStorage validates resources requests storage
+func validateRequestsStorage(requests corev1.ResourceList, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if _, ok := requests[corev1.ResourceStorage]; !ok {
+		allErrs = append(allErrs, field.Required(fldPath.Child("requests.storage").Key((string(corev1.ResourceStorage))), "PD or TiKV request storage must not be empty"))
```
Comment (aylei) [inline pkg/apis/pingcap/v1alpha1/validation/validation.go:178]:
```suggestion
		allErrs = append(allErrs, field.Required(fldPath.Child("requests.storage").Key((string(corev1.ResourceStorage))), "storage request must not be empty"))
```

We may invoke this method to verify other components and the error will include the field ref which is self-explanatory

**Comment by aylei** [comment]
Thanks! @shonge

**Pull request merged**